raysect/core/math/function/float/function3d/interpolate/discrete3dmesh.pyx

```diff
@@ -0,0 +1,180 @@
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+import numpy as np
+cimport numpy as np
+from raysect.core.math.function.float.function3d cimport Function3D
+from raysect.core.math.point cimport new_point3d
+cimport cython
+
+
+cdef class Discrete3DMesh(Function3D):
+    """
+    Discrete interpolator for data on a 3d ungridded tetrahedra mesh.
+
+    The mesh is specified as a set of 3D vertices supplied as an Nx3 numpy
+    array or a suitably sized sequence that can be converted to a numpy array.
+
+    The mesh tetrahedras are defined with a Mx4 array where the four values are
+    indices into the vertex array that specify the tetrahedra vertices. The
+    mesh must not contain overlapping tetrahedras. Supplying a mesh with
+    overlapping tetrahedras will result in undefined behaviour.
+
+    A data array of length M, containing a value for each tetrahedra, holds the
+    data to be interpolated across the mesh.
+
+    By default, requesting a point outside the bounds of the mesh will cause
+    a ValueError exception to be raised. If this is not desired the limit
+    attribute (default True) can be set to False. When set to False, a default
+    value will be returned for any point lying outside the mesh. The value
+    return can be specified by setting the default_value attribute (default is
+    0.0).
+
+    To optimise the lookup of tetrahedras, the interpolator builds an
+    acceleration structure (a KD-Tree) from the specified mesh data. Depending
+    on the size of the mesh, this can be quite slow to construct. If the user
+    wishes to interpolate a number of different data sets across the same mesh
+    - for example: temperature and density data that are both defined on the
+    same mesh - then the user can use the instance() method on an existing
+    interpolator to create a new interpolator. The new interpolator will shares
+    a copy of the internal acceleration data. The tetrahedra_data, limit and
+    default_value can be customised for the new instance. See instance(). This
+    will avoid the cost in memory and time of rebuilding an identical
+    acceleration structure.
+
+    :param ndarray vertex_coords: An array of vertex coordinates (x, y, z) with shape Nx3.
+    :param ndarray tetrahedras: An array of vertex indices defining the mesh tetrahedras, with shape Mx4.
+    :param ndarray tetrahedra_data: An array containing data for each tetrahedra of shape Mx1.
+    :param bool limit: Raise an exception outside mesh limits - True (default) or False.
+    :param float default_value: The value to return outside the mesh limits if limit is set to False.
+    """
+
+    def __init__(self, object vertex_coords not None, object tetrahedras not None, object tetrahedra_data not None, bint limit=True, double default_value=0.0):
```

Comment:
The keyword arguments of `KDTree3DCore` should be passed at initialisation of `Discrete3DMesh`. Probably as a dict, e.g. `kdtree_kwargs={'max_depth': 0, 'min_items': 1, 'hit_cost': 50.0, 'empty_bonus': 0.2}`.